# Introduction to Python

### Intro to Markdown in Jupyter

**1. Headings**

Headings are entered by prefacing text with one or more # (usually pronounced "hash") symbols followed by a space. If you know any HTML, then `# Title` is equivalent to a `<h1>Title</h1>`, `### Subtitle` is equivalent to `<h2>Subtitle</h2>`, etc. Note that you must put a space between the hashes and the text for it to be recognized as a heading.

**2. Text and paragraphs**

Regular text can be entered as normal. 
Blank lines between text indicate the start of a new paragraph; press `Enter` twice
at the end of a paragraph to leave such a blank line.
Other line breaks do not create a new paragraph and can be used
to avoid over-long lines of text or to separate sentences so they can be 
more easily reordered in the future.
 
**3. Emphasis**

The third cell (labeled "***Emphasis***") illustrates how to make text italicized or bold.  Double click on the cell (to view the source text).  
* Text can be italicized by surrounding it with single asterisks, like `*italicize text*`: *italicize text*.
* Similarly, text can be bolded by surrounding it with double asterisks, like `**bold text**`: **bold text**.
* Text can be both italicized and bolded by surrounding it with triple asterisks, like `***bold and italics***`: ***bold and italics***.

**4. Bulleted and Numbered Lists**

**Bulleted lists** can be created by starting a line with an asterisk and a space, followed by the text of the bulleted item. Bulleted sub-lists can be created by tab-indenting a bulleted list under a bulleted item. For example,

<pre>
* Item 1
    * Item 1.1
    * Item 1.2
    
*Item 2
</pre>

formats as
* Item 1
    * Item 1.1
    * Item 1.2
*Item 2

**Numbered lists** can be created by starting a line with a number, a period, and a space, followed by the item to be numbered. *Note that the first number you use in a sequence of numbered items will determine the starting point for numbering, but later values are ignored. It is a good habit to use 1 for every numbered item because then you can rearrange the items via cut-and-paste without worrying about upsetting the overall list numbering.*  Numbered sub-lists can be created by tab-indenting a numbered list under a numbered item. For example,

<pre>
1. Item 1
    1. Item 1.1
    1. Item 1.2
3. Item 2
</pre>
formats as
1. Item 1
    1. Item 1.1
    1. Item 1.2
3. Item 2

**5. Links and images**

**Links** are easily created by putting the link text in square brackets, followed by the link URL in parentheses, like:
```
[Example link](http://google.com)
```
which is rendered as 

[Example link](http://google.com)

**Images** are entered in a very similar way to URLs. Just put an exclamation point (!) before the square brackets. The text in the square brackets will be used as the "alt text", which is used by screen readers or displayed when hovering the mouse over the image. Images can also be inserted by dragging them from your file manager into a cell. After the Markdown code is inserted, you can edit the alt-text to something more meaningful.

![alt text for screen readers](./images/pandas-car-sales-csv.png "Text to show on mouseover")


**6. Other Markdown formatting**

Markdown has many other features that we will not cover, such as horizontal rules, block quotes, syntax highlighting, and tables.
A good reference for Markdown syntax is
[Markdown Guide (https://www.markdownguide.org/extended-syntax/)](https://www.markdownguide.org/extended-syntax/)


### Jupyter Magics

Code cells can also contain special instructions intended for JupyterLab itself, rather than the
Python kernel. These are called *magics*. For instance, to output your current directory, you can use the 
`%pwd` magic:

In [1]:
%pwd

'C:\\Users\\diwas\\OneDrive\\Desktop\\Data-Science-Specialization-with-Python\\Class'

```'~/data-science/class'```

You can use the "%cd" magic to change your directory (recall that~ is a shortcut for your home user directory):

In [ ]:
# cd ~

Changing directories is often useful to switch to a directory where data is stored.

You can get a list of magics and other information about Jupyter using the `%quickref` magic. The output of `%quickref` is long, so I am only including a screenshot of the top of its output.

In [ ]:
# %quickref

## Getting Started in Python

Python is an interpreted language, which means that when any Code cell in a Jupyter notebook is evaluated, the Python code will be executed. Any output or error messages will appear in a new output portion of the cell that will appear just after the input portion of the cell (that contains the Python code). 

### Hello World!

Let's start with the canonical example that appears in almost every programming language, which is to write "Hello World!" to the screen/output. 

To print output in Python, use the `print()` function, which is a built-in function in Python 3. Here a *function* is used to denote a *named* set of Python instructions that can be called on demand. Python functions are called by using the function name with parentheses after it. Functions can accept *arguments*, which are variables and values that the function can act on. The arguments for a function are put inside the parentheses, with commas separating different items. 

To output text using `print()`, put it as an argument inside single or double quotation marks. I will generally use single quotation marks because they do not require using the shift key to type.

In [ ]:
print('Hello World!')

Python's  `print()` function knows how to handle most standard data types. Unlike `printf` in C, explicit formatting instructions do not have to be given -- just pass the thing to be printed as an argument to `print`:

In [ ]:
print(32)

In [ ]:
print([1, 4, 9, 16])

To print multiple items, just pass all the items as arguments to the `print()` function, but separate the items by commas:

In [2]:
print('Hello', 'world!', 10 + 5, [1, 4, 9, 16])

Hello world! 15 [1, 4, 9, 16]


In [3]:
print("Diwash"+" "+"Sapkota")

Diwash Sapkota


Code cells may contain multiple Python statements. All statements in a cell will be run sequentially when the cell is run. The results of every `print` statement will be shown after the *input* part of the cell:

In [4]:
print('Hello ')
print('World!')

Hello 
World!


Some Python statements return results, and these will appear in a special *output* part of the cell:

In [ ]:
5 + 10

However, if a cell contains multiple statements, the output will be the result of the **last** statement:

In [ ]:
5 + 10
5 * 10

The last statement in a cell may produce no output. Then there is no output from the cell. Note that printed items are not outputs:

In [ ]:
5 + 10
print('hello')

There are several different ways to get the output of multiple statements. One easy way is just to print all the results you want to see:

In [ ]:
print(5 + 10)
print(5 * 10)

### Comments

Comments are text in your Python code that are ignored by the Python interpreter.
Documentation is usually used to convey what you intend
for your code to be doing -- which is not always what it is actually doing! 
Documentation is both for others who may need to read your code and for your future self, who
may need help remembering what you meant for a block of code to do. 

Anything that follows a # (hash) symbol is a comment:

In [ ]:
# It is important to use comments to document your thinking on big assignments

There is not really a multi-line comment in Python (like /* */ in C). One way to make a multi-line comment is to make a multi-line string that is not assigned to any variable.  Multi-line strings are delimited by triple-ticks (''') or triple-quotes("""). Here is an example with triple-quotes:

In [ ]:
""" This is a
multi-line comment,
okay?"""
print('The multi-line string above will not produce any output')
print('because it is not the last command in this cell.')

When making custom functions, a multi-line string directly after the function definition serves as the documentation string  (docstring) for that function.

### Python Variables and Types

Variables are named entities that store values for future reference. In Python, variable names consist of alphanumeric characters (a-z, A-Z, 0-9) and underscores, but they cannot start with a number. Variable names are case-sensitive.  Detailed descriptions of naming conventions for Python are given in [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/]). Generally, we will follow these conventions:
* Names should be descriptive, except when used as a simple index.
* Variable and function names should be lowercase, with underscores separating different words.
* Variables should not duplicate the names of modules or Python functions/methods.

Variables are created by assigning a value to a valid variable name:

In [ ]:
x = 10
temp = 50.5
temp3 = 'Diwash Sapkota' 

In [ ]:
type(temp3)

Python uses implicit typing, which means that you do not have to define a variable's type (i.e., what type of data it holds). The type is determined by the interpreter at the time of assignment. You can determine a variable's type by passing the variable as the argument of the `type()` function:

In [ ]:
type(x)

The type of a variable can change whenever new data is assigned to it:

In [ ]:
x = 10.5

In [ ]:
type(x)

In [ ]:
x = 'Hello World!'

In [ ]:
type(x)

When a variable is passed as an argument to the `print()` function, the value of the variable is printed:

In [ ]:
print(x)

Python knows how to perform many different operations on different data types and will usually do the *right thing* based on the type of the variable:

In [ ]:
a = 3
b = 4
a + b

In [ ]:
print(a * b)

We often want to combine some fixed text and some variable output. To do this, we will use *f*-strings, which were added to Python in version 3.6. An *f*-string is a special string that is created by prefixing the delimiter with the letter `f`.  Any part of an *f*-string contained within curly braces {} will be evaluated before the string is used. Thus, if we wanted to write more about the product of `a` and `b`, we could do it like this:

In [ ]:
print(f'The product of a and b is {a*b}')

We can have multiple substitutions in a single *f*-string:

In [ ]:
print(f'The product of a and b is {a*b}')

```{warning}
Exponentiation is performed using two asterisks between the number or variable to be operated on and the exponent. The carat operator (^) is reserved for bitwise exclusive-or (XOR).
```

In [ ]:
2 ** 3

We will often need to add to an existing value, and Python provides a shortcut notation for this (similar to C). The notation `+=` indicates to add to the variable on the left-hand side of the expression. For instance, here is the usual way we write incrementing a counter:

In [ ]:
counter = 5
# counter = (counter * 5)
counter *= 10
print(counter)

Most operators in Python try to do the most logical operation for their input types. Therefore, if we try to add strings with +, the operator actually performs string concatenation:

In [ ]:
a = 'Hello '
b = 'World'
print(a + b)

In [ ]:
a = 3
b = 4.1
print(a + b, type(a + b))

However, Python cannot read minds, and its rules may create unexpected results:

In [ ]:
a = '3'
b = '4'
print(a + b)
print(int(a) + int(b))

Note that in the first case, the variables are strings, so the + operator performs string concatenation. <mark> This can be a subtle problem when we load data from files because items that look like numbers may actually be loaded as strings. </mark>

The multiplication operator performs repetition on a string:

In [ ]:
a = '3'
c = 5
print(a * c)
print(int(a) * c)

### Basic Data Types

Python has many data types built-in, and we have already seen a few of these: *int*, *float*, and *str* (string). We will also see few other built-in data types

Boolean values are stored in *bool* variables that take on either `True` or `False`. Note that exact capitalization is required for these values.

In [ ]:
a = True
print(a)

We can check if two variables have the same values using `==`:

In [1]:
b = 2
c = 3
C = 4
print(c)
# print(b == c)

3


The result of a comparison (using ==) can be stored in variable (using =):

In [ ]:
d = (b == c)
d
# type(d)

Not equals is written as `!=`. Other comparisons are written in the usual ways (<, >, <=, >=).

In [ ]:
print(b != c, b <= c)


Python has a variety of different *sequence* types for containing an ordered  collection of values. Values can be retrieved by index from a sequence-type variable by putting the index number in square brackets after the variable name.

```{warning}
Like C, C++, and JavaScript, indexing in Python starts at 0. This means that in a sequence of $n$ items, the first item is at index 0, and the last item is at index $n-1$. MATLAB and many textbooks use indexing starting at 1.
```

The *list* type is a **mutable** container of values. *Mutable* means the values in the list can be changed. Lists are delimited by square brackets, and list items are separated by commas. 

In [9]:
my_list = ['dogs', 'cats', 3, 7.0]
print(my_list)

my_list2 = [[1, 3, 5], [2, 4, 6], 'diwash', 'kathmandu', 4.0]
my_list2

['dogs', 'cats', 3, 7.0]


[[1, 3, 5], [2, 4, 6], 'diwash', 'kathmandu', 4.0]

In [6]:
print(my_list[1])

cats


In [7]:
print(my_list[-1])

7.0


In [1]:
my_list3 = ['mango', 'apple']
my_list3.sort()
print(my_list3)

my_list4 = sorted(my_list3, reverse=True)
print(my_list4)

my_list4.append('orange')
print(my_list4)

# my_list4.pop(-1)
# my_list4

my_list4[0] = 'banana'
print(my_list4)

['apple', 'mango']
['mango', 'apple']
['mango', 'apple', 'orange']
['banana', 'apple', 'orange']


In [25]:
# for i in range(my_list2):
#     # print(i)

Because lists are mutable, values can be updated:

In [11]:
print(my_list)
my_list[0] = 'birds'
print(my_list)

['dogs', 'cats', 3, 7.0]
['birds', 'cats', 3, 7.0]


A *tuple* is an **immutable** sequence type. The values in a tuple cannot be changed. Tuples are often used to contain multiple values returned from a function. Tuples are delimited by parentheses: (). To create a tuple containing only one value, include a comma after the value. As with lists, tuples can contain a variety of values. Trying to change a value in an immutable type results in an error:

In [12]:
tuple1 = (1, 4, 'nine')
tuple2 = (16,)

In [13]:
print(tuple1[2])

nine


Trying to change a value in an immutable type results in an error:

In [29]:
tuple1[0] = 5

TypeError: 'tuple' object does not support item assignment

The *range* type is an **immutable** sequence of numbers that is usually used in looping (especially `for` loops). A range object can be created using the `range()` function in the following ways:
* `range(stop)` creates a sequence of *stop* values starting at 0. Thus, the values are 0, 1, 2, $\ldots,$ *stop*-1. Note that the Python convention is that the *stop* value is not included in the range. 
* `range(start, stop)` creates a sequence of values that starts at *start* and ends at *stop*-1. 
* `range(start, stop, step)` creates a sequence of values that starts at *start*, increments by *step*, and ends at *stop*-1.

You can iterate over ranges using a `for`...`in` statement, which is discussed later.

Python also contains one *mapping* type:
* *dict* is a dictionary object that provides a map between key-value pairs. 
As with lists and tuples, keys and values can be any data types, but the keys must be unique.
Dictionaries are delimited by curly braces: { }. Each entry in a dictionary is written in the form `key:value`, and different key-value pairs are separated by commas. The value for a particular key can be retrieved by putting the key in square brackets after the dictionary variable's name.

In [2]:
squares = {1: 1, 2: 4, 3: 9, 4: 16}
print(squares[3])

9


In [3]:
keys = ["name", "age", "city"]
values = ["Alice", 25, "New York"]
user_info = dict(zip(keys, values))
print(user_info)

{'name': 'Alice', 'age': 25, 'city': 'New York'}


In [5]:
misc = {
    'cats': 0, 
    'dogs': 1, 
    3: 'test'
}
print(misc['cats'])
print(misc[3])

0
test


Python variables are actually much more powerful than variables in many languages because they are actually *objects*. Python is an object-oriented programming (OOP) language. Fundamentals about **objects** and **classes**:
* *Objects* are special data types that have *methods* associated with them to work on those objects. Methods are similar to functions, except they are specialized to the *objects* to which they belong. Methods are called by giving the variable/object name, adding a period, specifying the method name, and then adding parentheses, with any arguments provided in parentheses. 
* A *class* is a template for an object that defines how an object stores information and defines its methods.

For instance, since *list* is a mutable type, a list can be sorted in place. The *list* object defines a sorting method to achieve this:

In [14]:
new_list = [5, 7, 1, 3, 13]
new_list.sort()
print(new_list)

[1, 3, 5, 7, 13]


In [16]:
# For sorting, list must have same data type
my_list = [1, 4, 3, 'hello']
my_list.sort()
print(my_list)

TypeError: '<' not supported between instances of 'str' and 'int'

For dictionaries, we often need to retrieve the keys or values. We can do this using methods provided by the dictionary type:

In [6]:
misc = {'cats': 0, 'dogs': 1, 3: 'test'}
print(misc.keys())

dict_keys(['cats', 'dogs', 3])


In [18]:
print(misc.values())

dict_values([0, 1, 'test'])


The `keys` and `values` methods return special objects of type `dict_keys` and `dict_values`, respectively, but for our purposes, we can treat these like lists.

### Copying Variables

```{warning}
You have to be careful when copying variables in Python, or you may get some unexpected results.
```

Suppose you have a variable c that contains a list, and you want to copy it to a variable d:

In [ ]:
c = [0, 1]
d = c
print(c, d)

c[1] = 5
print(c, d)

Everything works as expected. Now, let's change the value in position 1 in `d` and print out both `c` and `d`:

In [ ]:
d[1] = 2
print('d=', d)
print('c=', c)

Changing the value of `d[1]` also changed the value of `c[1]`! This is because when `c` is made, `c` is a variable that points to the list `[0, 1]`. When we set `d=c`, Python sets the variable `d` to point to the **same** list `[0,1]`.

To make a copy of the list `c`, we can use the list's `copy()` method to create a new list with the same contents as `c`:

In [ ]:
c=[1,0]
d = c.copy()
print(c, d)
c[1] = 4
print(c, d)

We can check if two variables point to the same data using Python's `is` command:

In [ ]:
c is d

If two variables point to the same data, then they will have the same values:

In [ ]:
c == d

However, the opposite is not true. If we make a new copy of `c`, it will point to a different list object, but the contents will be the same:

In [ ]:
f = c.copy()
print(f is c, f == c)

### Indentation and Line Breaks in Python

Line breaks in Python are usually used to distinguish different Python statements. Exceptions to this include:
* When a statement includes arguments in parentheses, line breaks can be used within the parentheses to improve readability. The statement will not terminate until the closing parenthesis, which should be followed by a line break.
* As previously introduced, strings that include line breaks can be created using triple quotes.
* If a line ends in a backslash `\`, it will be interpreted as continuing onto the next line.
* Multiple statements can be put onto a single line by separating them by semicolons (;), but this is not encouraged.


In [ ]:
print('Hello', 'Michael!')

In [ ]:
a = 2 + 3
print(a)

In [ ]:
print(
    'Goodbye, John!'
)

In [ ]:
print(
    '''Goodbye, 
John!'''
)

In [ ]:
# Here is an example of using a backslash to break up a line into smaller parts:
A = 1 + 2 + 3 \
  + 4 + 5 + 6
A

Programming languages usually have some convention to indicate which statements belong together. For instance, if a statement starts a loop (such as `for` or `while`), there must be a way to indicate which of the following statements should be executed in each iteration of the loop and which should be executed after loop iteration. In many languages, such as C, C++, Java, and Javascript, code blocks are surrounded by curly braces: { }.

In Python, indentation is used to deliminate code blocks. The languages mentioned above use indentation as a *convention* that indicates meaning to humans working on code. Python uses indentation to *define* code blocks and convey meaning to the Python interpreter.

Either tabs or spaces can be used to denote code blocks, but the PEP-8 standard is to use 4 spaces per indent level. Jupyter inserts 4 spaces when `Tab` is pressed in a code block.[^tabspaces]

[^tabspaces]:  Some text editors can be set up to insert a prescribed number of spaces when tab is used or convert tabs to spaces upon saving.



### Loops and Conditionals

#### `for`... `in` Statements
In data science, we often need to either iterate over data or carry out iterations of a simulation. For both purposes, we will usually rely on Python's `for`...`in` statement. This is a type of *compound* statement. Compound statements consist of a *header* and a *suite* (Python's terminology) or *body*. The header always ends in a colon, and the *suite* is one more statements that are run consecutively according to conditions in the header. For our purposes, we will always create the suite as a set of statements that are indented one more level than the corresponding header, with each statement on a new line. 

The `for`...`in` compound statement takes a variable after the `for` keyword. This variable will hold the current iteration value. After the `in` keyword  there needs to be an *iterable* object, which is any object that can be iterated over. The typical one we will use is the `range` object that we have just introduced:

In [19]:
for i in range(10):
    print(i)
    print('This is in the loop')
print()
print('This is after the loop is run')

0
This is in the loop
1
This is in the loop
2
This is in the loop
3
This is in the loop
4
This is in the loop
5
This is in the loop
6
This is in the loop
7
This is in the loop
8
This is in the loop
9
This is in the loop

This is after the loop is run


In [56]:
for j in range(2, 4):
    print(j)

2
3


In [59]:
for k in range(2, 11, 2):
    print(k)

2
4
6
8
10


Another example of an iterable object is a list:

In [21]:
fruit = ['apples', 'bananas', 'cherries']
for kind in fruit:
    print(kind)

print()
print('Following is the above list reversed')
fruit = ['apples', 'bananas', 'cherries']
for i in reversed(fruit):
    print(i)

apples
bananas
cherries

Following is the above list reversed
cherries
bananas
apples


When iterating over objects like a list, it is often helpful to keep track of the iteration index. One easy way to do this is to use the `enumerate()` function that returns a tuple of the current iteration index and the item:

In [68]:
for i, kind in enumerate(fruit):
    print(i)
    print(kind)
    print()

0
apples

1
bananas

2
cherries



In the previous example, I purposefully split the printing into different lines to show a multi-statement suite.

We can *nest* 'for' statements, which means that one `for` statement is inside of another `for` statement. For each iteration of the outer loop, the inner loop will run through all its iterations:

In [7]:
for i in range(4):
    print(f'Iteration {i}')
    for j in range(3):
        print(i, j)
    print()

Iteration 0
0 0
0 1
0 2

Iteration 1
1 0
1 1
1 2

Iteration 2
2 0
2 1
2 2

Iteration 3
3 0
3 1
3 2



In [28]:
students = {
        'name' : 'John',
        'class' : 9,
        'address' : 'Kathmandu'
 }

In [23]:
for item in students.keys():
    print(item)
    
print()
for item in students.values():
    print(item)
print()
for i, j in students.items():
    print(i, j)

name
class
address

John
9
Kathmandu

name John
class 9
address Kathmandu


In [30]:
deletedval = students.pop('class')
# print(deletedval, students)

9

#### `if` Statements

The `if` statement is used to run code conditionally. It will often be the case that `if` statements are used inside  loop or other conditional statements. For instance, to print out only the even numbers from 1 to 10, we can use an `if` inside a `for` statement:

In [73]:
for i in range(1, 11):
    if i % 2 == 0:
        print(i)
    else:
        print('Not even')

Not even
2
Not even
4
Not even
6
Not even
8
Not even
10


**If Else Statements**

 More generally, an `if` statement may also have `elif` and `else` clauses.  *Elif* is short for "else if", and these headers act like `if` headers but will be evaluated only if the above `if` or `elif` headers did not have their conditions satisfied. There can only be one `else` clause, and its suite will be executed if the `if` and `elif` clauses did not have their conditions satisfied.  Note that the `elif` and `else` headers must be at the same indentation level as the corresponding `if` header, and these headers must also end with a colon (:). The statements that belong to the suites for these headers  go on separate lines below the header and are indented to one greater level.
 
As a simple example, suppose we want to identify the even numbers from 1 to 10, but if a number is NOT even, we wish to determine if it is divisible by 3. Otherwise, we just want to print an asterisk:

In [75]:
for i in range(1, 11):
    if i % 2 == 0:
        print(f'{i} is even')
    elif i % 3 == 0:
        print(f'{i} is not even but divides by 3')
    else:
        print('*')

*
2 is even
3 is not even but divides by 3
4 is even
*
6 is even
*
8 is even
9 is not even but divides by 3
10 is even


Note how indentation changes meaning in the following two examples, which differ only by one Tab:

In [22]:
a = 2
b = 3
if a == 2:
    print('a=2')
if b == 2:
    print('b=2')
    print(a, b)

a=2


In [23]:
a = 2
b = 3
if a == 2:
    print('a=2')
if b == 2:
    print('b=2')
print(a, b)

a=2
2 3


We will generally iterate over dictionaries by iterating over their keys:

In [24]:
squares = {1: 1, 
           2: 4, 
           3: 9, 
           4: 16, 
           5:25
          }
for i in squares.keys():
    # print(f'{i}**2 = {squares[i]}')
    print(f'{i ** 3}')

1
8
27
64
125


In [ ]:
squares[5]

In [25]:
students = {
        'name': 'Diwash Sapkota',
        'age': 20,
        'major': 'Computer Science'
}

In [26]:
for i in students:
    print(students[i])

Diwash Sapkota
20
Computer Science


#### `while` Statements

A `while` loop combines looping and a conditional statement for determining whether looping should continue. It operates similarly to `while` loops in other programming languages. The loop will continue as long as the condition specified in the `while` statement is satisfied:

In [1]:
i=2
while i < 10:
    # conditions to run
    i = i + 3
    print(i)

5
8
11


In [3]:
number = int(input('Enter any number between 100 and 500 '))
# number greater than 100 and less than 500
while number < 100 or number > 500:
    print('Incorrect number, Please enter correct number:')
    number = int(input('Enter a Number between 100 and 500 '))
else:
    print("Given Number is correct", number)

Enter any number between 100 and 500  600


Incorrect number, Please enter correct number:


Enter a Number between 100 and 500  700


Incorrect number, Please enter correct number:


Enter a Number between 100 and 500  180


Given Number is correct 180


In [4]:
# Iterate over a list using while loop
numbers = [1, 2, 4, 5, 7]
size = len(numbers)
i = 0
while i < size:
    print(numbers[i])
    i = i + 2

1
4
7


In [9]:
# Initialize a counter
count = 1
# Use a while loop to print numbers from 1 to 10
while count <= 10:
    if count == 5:
        print("Skipping 5 (using continue)")
        count += 1
        continue  # Skip the rest of the loop body for this iteration
    print(count)
    if count == 8:
        print("Exiting loop at 8 (using break)")
        break  # Exit the loop when count reaches 8
    count += 1  # Increment the counter
print("Loop finished!")

1
2
3
4
Skipping 5 (using continue)
6
7
8
Exiting loop at 8 (using break)
Loop finished!


In [10]:
i = 10

while i >= 1:
    print(i)
    i = i - 1

10
9
8
7
6
5
4
3
2
1


### Functions

We have already used built-in functions, like `print()` and `type()`. 
Sometimes, we will also need to create new functions. As with the built-in functions, 
user-defined functions can take arguments and can return values. Functions are declared 
using the Python keyword ```def``` followed by the function name and then parentheses. 
Inside the parenthesis, list any variables that will receive arguments, separated by commas.  Here is a simple
example:

In [11]:
def say_hello(name):
    print(f'Hello, {name}!')

The function name and parameter list are called the *function signature*. User-defined functions are called in the same way as built-in functions:

In [12]:
say_hello('Diwash')

Hello, Diwash!


Functions can also return values by placing them after the Python `return` keyword. If multiple values are to be returned, they should be separated by commas. If more than one value is returned, it will be returned as a tuple:

In [14]:
def square_and_cube(x):
    return x ** 2, x ** 3

In [16]:
square_and_cube(4)

(16, 64)

In [18]:
def odd_even_check(x):
    if (x % 2) == 0:
        print("Even")
    else:
        print("Odd")

In [19]:
odd_even_check(5)

Odd


When storing returned values from a function into multiple variables, you do not have to explicitly use the parentheses around the tuple of variables. You can just separate the variables by commas, as shown below:

In [ ]:
four2, four3 = square_and_cube(4)

In [ ]:
print(four2, four3)

You can also provide a docstring for a function as 
a string that directly follows the function definition. For example, let's define a function
that returns the squared error between its inputs:

In [20]:
def squared_error(x, y):
    """
    Returns the squared error 
    (the squared difference) of the arguments
    """
    return (x - y) ** 2

In [23]:
z = squared_error(5, 7)
z

4

In [ ]:
squared_error(2, 3)

Python allows you to specify default values for function arguments that will be used if the user does not pass a value for the argument. When defining a function, specify the default value for an argument by writing the parameter name, followed by an equal sign, followed by the default value in the function signature.

Let's make a new version of `squared_error` function that sets the default value of `y` to 0:

In [25]:
def squared_error2(x, y, z=0):
    """
    Returns the squared error between the two arguments, with a default value of 0
    for the second argument
    """
    print(f'x = {x} y = {y} z = {z}')
    return (x - y + z) ** 2

In [26]:
squared_error2(1, 2, 3)
squared_error2(y = 2, x = 5, z = 6)

x = 1 y = 2 z = 3
x = 5 y = 2 z = 6


81

If the user passes a value, the default value is not used:

In [27]:
# function with default argument
def show_employee(name, salary=45000):
    print("Name:", name, "salary:", salary)

show_employee("John", 120000)
show_employee("Matthew")

Name: John salary: 120000
Name: Matthew salary: 45000


In [ ]:
squared_error2(2, 3)

Note that we can use the names of the parameters (instead of parameter order) to pass values to those parameters. This is very commonly used in functions that have lots of parameters that are optional.

In [ ]:
squared_error2(y=2, x=3)

Parameters can be passed using a mix of order and parameter names. However, any parameters passed by order must come before those passed by name.

We will see in the next section how to get help on a function.

### Getting Help and Completion

Python has built-in help for almost every function and object. This help can be retrieved in several ways. For instance, consider the built-in `sum` command. Here are several ways to get help in Jupyter:

In [ ]:
help(sum)

In [ ]:
?sum

In [ ]:
?append

For user-defined functions, 'help' will display the docstring you wrote. The following assumes that you have defined the function `squared_errors` from Section {ref}`python-intro:functions`:

In [ ]:
?squared_error

If a function's Python source code is available, it can be retrieved in Jupyter using `??`:

In [ ]:
??squared_error

Now, let's look at the help for a variable. 

In [ ]:
squares = {1: 1, 2: 4, 3: 9, 4: 16}

In [ ]:
?squares

You should also try `help(x)`, but I have omitted that because it provides help for every method of the `dict` object, which results in a lot of output.

You can also try `help()` with no argument to get an interactive help session.

Jupyter also provides many features to help you during programming. Assuming you have run the command defining `squares` above, try the following in a new Jupyter notebook cell:
1. Type `sum(`. When you type the open parenthesis, Jupyter should automatically insert a pair of parentheses. 
2. Press `shift-Tab`. You should see the call signature and doc string for the `sum` function in a pop-over box. You can press the `Esc` key to close the pop-over box. 
3. Type `sq` and press `Tab`. You should see a list of variables and functions that begin with `sq`. Use the cursor keys to scroll to `squares` and press `Enter` to insert it without having to type the full name of the squares dictionary.
4. Let's sum the values in the `squares` dictionary. Type a period and then press `Tab` again to see a list of methods for a `dict` object. Select `values` using the keyboard or mouse.
5. Don't forget that we need parentheses to call the `values` method. Press `(` and a pair of parentheses should appear.
6. Press `shift-Enter` to run the cell.

In [ ]:
sum(squares.values())

### Python Modules and Namespaces

Python has many useful modules that extend Python's basic functionality. Some of these are included with the base Python distribution, and many others are included in the Anaconda distribution. Many more can be installed over the Internet. 

To use a module, you must import it into your Python working environment. The most basic way to do this is to type `import` followed by the name of the module to be imported:

In [36]:
import numpy
import pandas

Here we have imported NumPy (usually pronounced "Numb Pie"), one of the most important Python modules for working with numerical functions and arrays. When a module is imported, its functions and classes will be available in Python, but they are imported into their own *namespace*. To access something in a different namespace, type the name of the namespace, followed by a period, followed by the name of the thing you are trying to access.

For instance, the value of $\pi$ is a constant object named `pi` in NumPy. Now that we have imported NumPy, we can access that value:

In [37]:
print(numpy.pi)

3.141592653589793


NumPy has many typical mathematical functions, which we can call using the `numpy` namespace:

In [38]:
numpy.sin(numpy.pi / 4)

0.7071067811865476

In [ ]:
numpy.sqrt(2) / 2

We can control the namespace into which the contents of a module is imported. Because many modules, like NumPy, are commonly used, the community often uses community-standardized namespaces that are shorter than the full module name. To do this, type `import`, followed by the module name, followed by the `as` keyword, followed by the desired namespace. 

For NumPy, the data science community typically uses `np`, so the import statement is as follows:

In [32]:
import numpy as np
import pandas as pd

In [33]:
print(np.pi)

3.141592653589793


```{warning}
It is possible to import the contents of a module into the *global* namespace, which means that the namespace does not have to be specified before each function, class, or object. However, this practice is strongly discouraged because it often results in conflicts. For instance, both Matplotlib (a plotting module) and SymPy (a symbolic algebra module) have a `plot` function. If you were to import both `matplotlib` and `sympy` into the global namespace, you could not be sure which `plot` you were calling, unless you kept track of which module was imported last.
```


Importing into namespaces is a good practice. However, on occasion, it may be helpful to import just a single function from a module, and in this case, it is reasonable to import it into the global namespace if we can be confident that there will not be any collisions. An example follows:

In [28]:
from scipy.special import factorial

In [30]:
factorial(5)

120.0

Why is factorial returning a float? Let's check the docstring:

In [32]:
?factorial

By inspecting the docstring, we can see that if we want the exact value, we need to set the parameter `exact` to True. This is typically done using the parameter name because anyone reading the function call will understand what the value of `True` is being used for:

In [ ]:
factorial(10, exact=True)

#### NumPy Arrays

NumPy provides a `numpy.ndarray` container for holding one-dimensional or multi-dimensional collections of numbers. Most engineers will have some familiarity with vectors or matrices.  For our purposes, we will use the following definitions of these mathematical objects:

````{card}
DEFINITION
^^^
```{glossary}
vector
    A one-dimensional, ordered list of numbers that has an accompanying notion of magnitude of a vector and distance between two vectors. 
```
````

Vectors are usually shown enclosed in square brackets. When writing as mathematics notation, we will use bold, lowercase letters to denote a vector. In mathematics, a vector may either be a *column vector* or a *row vector*. If not otherwise specified, *vector* will refer to a column vector; for instance, here is an example of defining a vector of the first five counting numbers:

$$
\mathbf{x} =
\begin{bmatrix}
1 \\
2 \\
3 \\
4 \\
5 \\
\end{bmatrix}.
%\mathbf{x} = \left[ 1,2,3,4,5 \right].
$$

NumPy vectors do not have any notion of direction (column or row) but are usually displayed as a row and are interpreted as a row by some NumPy operations. To make a NumPy vector, call `np.array()` with the vector elements enclosed in square brackets and separated by commas:



In [34]:
x = np.array([[1, 2, 3, 4, 5], [1, 2, 3, 4, 5]])
x

array([[1, 2, 3, 4, 5],
       [1, 2, 3, 4, 5]])

Arrays extend the concept of a vector to multiple dimensions:

````{card}
DEFINITION
^^^
```{glossary} 
array
    A multi-dimensional table of numbers that supports a standard set of operations, including multiplication of matrices.  


````{card}
DEFINITION
^^^
```{glossary}
matrix
    An alternate term for a two-dimensional array.
```
````

Note that an array can have one dimension, in which case it is a vector; in such cases, I will always refer to it as a vector. I will use *array* for higher dimensions, even though we will generally only consider two-dimensional arrays. Mathematically, two-dimensional arrays are shown as a table of numbers, organized into rows and columns, and enclosed in large square brackets.  For example, the array below has rows that contain consecutive powers (1, 2, and 3) of the first five elements:

$$
\begin{bmatrix}
1 &2 &3 &4 &5 \\
1 & 4 & 9 & 16 & 25 \\
1 & 8 & 27 & 64 & 125 \\
\end{bmatrix}
$$


Arrays are very helpful for doing basic data manipulation and storing numerical values for simulations, so we review some basics here.

In Python, arrays can be created using NumPy's `np.array()` function.
Two-dimensional arrays can be created by passing a list  whose contents are equal-length lists of numbers. Each of the interior lists of numbers represents one row of the array. This will be most clear with an example:

In [37]:
B = np.array(
  [[1, 2, 3, 4],
   [8, 7, 6, 5]]
)

B

array([[1, 2, 3, 4],
       [8, 7, 6, 5]])

NumPy arrays offer many advantages over lists. Two primary ones are:
1. It is easy to perform numerical operations on the elements of NumPy arrays.
2. NumPy arrays provide a variety of built-in methods that we will find useful.

To illustrate these, consider the array $B$ defined above and a Python `list` $L$ containing the same numbers:

In [38]:
L = [1, 2, 3, 4]

The NumPy array makes it easy to multiply all the elements in the array by a value:

In [39]:
B * 5

array([[ 5, 10, 15, 20],
       [40, 35, 30, 25]])

Compare the result with the effect of the multiplication operator on a list:

In [40]:
L * 5

[1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4]

An example of a built-in method is `sum()`:

In [41]:
B.sum()

36

The Python `list` object does not have a built-in `sum()` method, but Python does offer a general `sum` function:

In [42]:
L.sum()

AttributeError: 'list' object has no attribute 'sum'

In [43]:
sum(L)

10

We will introduce other methods of the `array` type and other NumPy functions that work on and/or return arrays as we introduce more data-science techniques. 

### Numerical Errors and Rounding

Python is not able to store all numbers exactly internally, and mathematical operations on numbers can also produce numerical errors. Let's illustrate this with a simple example:

In [ ]:
3*0.1

Based on this, you might conclude that the internal representation of 0.1 is actually a little larger than 0.1, and answers involving sums or multiples of 0.1 might sometimes come out a little too large. However, the truth is much more subtle, as shown by this example:

In [ ]:
sum=0
for i in range(8):
  sum+=0.1
  print(sum)

These are *computational errors*, and we will occasionally round the outputs of functions to reveal the true values. We can use the NumPy function `np.round()`, which takes two arguments: 1) the number or array to be rounded, and 2) the number of digits of precision to preserve:

In [ ]:
sum=0
for i in range(8):
  sum+=0.1
  print(np.round(sum, 10) )

I will generally round to 10 digits of precision, as that is usually enough to remove any computational error while preserving any decimal values that are not caused by computational error.

### Writing Big Numbers in Python

We will be building simulations in Python that require looping thousands or millions of times. Thus, we will often be writing a `range` that has an argument with many zeros. The `range` function will not take a `float` value, and numbers written in scientific notation (like `1e6`) will be treated as floats. This results in using integers that are very hard to read, like 10000000. We can't use commas in the numbers because that would create a tuple:


In [ ]:
10, 000, 000

Fortunately, Python provides a simple way to make large numbers like these more readable. Instead of using commas as a delimiter between every third digit, use underscores (\_). Doing this makes it much easier to interpret large numbers, like ten million:

In [ ]:
10_000_000

### Additional Resources

For those who want to learn more about Python, the following resources are recommended:
* [*A Whirlwind Tour of Python* (https://jakevdp.github.io/WhirlwindTourOfPython/)](https://jakevdp.github.io/WhirlwindTourOfPython/) by Jake VanderPlas is a free eBook that covers all the major syntax and features of Python.
* [Learn Python for Free (https://scrimba.com/learn/python)](https://scrimba.com/learn/python) is a free 5-hour online introduction to Python (signup required).
* The Python documentation includes a [Python Tutorial (https://docs.python.org/3/tutorial/)](https://docs.python.org/3/tutorial/).

## Review

**Self-Assessment:**

The following questions can be used to check your understanding of the material covered in this section:

In [50]:
# from jupyterquiz import display_quiz

# git_url='https://raw.githubusercontent.com/jmshea/Foundations-of-Data-Science-with-Python/main/'
# file = '01-intro/questions/6-intro-jupyter-and-python.json'
# display_quiz(git_url+file, colors='fdsp')
# #display_quiz('questions/intro-jupyter-and-python.json', colors='fdsp')